p2p/src/transport/mconnection.rs

```diff
@@ -0,0 +1,520 @@
+impl MConnection {
+    /// Opens a TCP connection to a remote host and establishes a secret connection
+    /// [`SecretConnection`].
+    fn connect(
+        addr: &SocketAddr,
+        private_key: &ed25519::Keypair,
+        protocol_version: Version,
+    ) -> Result<Self> {
+        let stream = TcpStream::connect(addr)?;
+
+        let stream_clone = stream.try_clone()?;
+        let local_addr = stream.local_addr()?;
+        let peer_addr = stream.peer_addr()?;
+
+        let public_key = private_key.public;
+        let secret_connection = SecretConnection::new(stream, private_key, protocol_version)?;
+
+        let (sender, loop_receiver) = flume::bounded(0);
+        let virtual_streams = Arc::new(RwLock::new(HashMap::new()));
+
+        let virtual_streams_clone = virtual_streams.clone();
+        std::thread::spawn(move || {
+            MConnection::rw_loop(secret_connection, loop_receiver, virtual_streams_clone)
+        });
+
+        Ok(Self {
+            public_key,
+            stream: stream_clone,
+            local_addr,
+            peer_addr,
+            sender,
+            virtual_streams,
+        })
+    }
+
+    /// Opens a TCP connection to a remote host with a timeout and establishes a secret connection
+    /// [`SecretConnection`].
+    fn connect_timeout(
+        addr: &SocketAddr,
+        timeout: Duration,
+        private_key: &ed25519::Keypair,
+        protocol_version: Version,
+    ) -> Result<Self> {
+        let stream = TcpStream::connect_timeout(addr, timeout)?;
+
+        let stream_clone = stream.try_clone()?;
+        let local_addr = stream.local_addr()?;
+        let peer_addr = stream.peer_addr()?;
+
+        let public_key = private_key.public;
+        let secret_connection = SecretConnection::new(stream, private_key, protocol_version)?;
+
+        let (sender, loop_receiver) = flume::bounded(0);
+        let virtual_streams = Arc::new(RwLock::new(HashMap::new()));
+
+        let virtual_streams_clone = virtual_streams.clone();
+        std::thread::spawn(move || {
+            MConnection::rw_loop(secret_connection, loop_receiver, virtual_streams_clone)
+        });
+
+        Ok(Self {
+            public_key,
+            stream: stream_clone,
+            local_addr,
+            peer_addr,
+            sender,
+            virtual_streams,
+        })
+    }
+
+    fn rw_loop(
+        mut secret_connection: SecretConnection<TcpStream>,
+        rx: Receiver<PacketMsg>,
+        virtual_streams: Arc<RwLock<HashMap<StreamId, Sender<PacketMsg>>>>,
+    ) {
+        loop {
+            // If any stream is trying to send a message, go for it.
+            if let Ok(msg) = rx.try_recv() {
+                let mut buf = Vec::new();
+                msg.encode_length_delimited(&mut buf)
+                    .expect("encode to always succeed");
+                if let Err(e) = secret_connection.write_all(&buf) {
+                    println!("can't write to connection: {}. exiting...", e);
```

Comment:
@thanethomson @xla what is the convention around logging inside the "command loop" running in a separate thread? how do I inform the caller about any errors? Should I return a channel and write any errors to it?

Reply:
Is this a terminal condition for the entire transport?

Reply:
No, just for that particular connection.

Reply:
This should be surfaced through one of the public methods of `Connection` so whenever the owner tries to interact with it, by opening a stream or reading from a stream.